# Nama       : Rismayanti 
# NIM        : 2109106025
# Posttest 4 : Kecerdasan Buatan

## Mengimport Library dan Membaca File CSV

In [32]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

dataset_split = pd.read_csv("healthcare-dataset-stroke-data.csv")

## Menampilkan Informasi dari Dataset

In [33]:
dataset_split.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5110 entries, 0 to 5109
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 5110 non-null   int64  
 1   gender             5110 non-null   object 
 2   age                5110 non-null   float64
 3   hypertension       5110 non-null   int64  
 4   heart_disease      5110 non-null   int64  
 5   ever_married       5110 non-null   object 
 6   work_type          5110 non-null   object 
 7   Residence_type     5110 non-null   object 
 8   avg_glucose_level  5110 non-null   float64
 9   bmi                4909 non-null   float64
 10  smoking_status     5110 non-null   object 
 11  stroke             5110 non-null   int64  
dtypes: float64(3), int64(4), object(5)
memory usage: 479.2+ KB


## Menampilkan Jumlah Record dan Atribut

In [34]:
dataset_split.shape

(5110, 12)

In [35]:
dataset_split.isna().sum()

id                     0
gender                 0
age                    0
hypertension           0
heart_disease          0
ever_married           0
work_type              0
Residence_type         0
avg_glucose_level      0
bmi                  201
smoking_status         0
stroke                 0
dtype: int64

## Menampilkan 10 Data Pertama
### Dataset merupakan jenis dataset categorical

In [36]:
dataset_split.head(10)

id  gender   age  hypertension  heart_disease ever_married  \
0   9046    Male  67.0             0              1          Yes   
1  51676  Female  61.0             0              0          Yes   
2  31112    Male  80.0             0              1          Yes   
3  60182  Female  49.0             0              0          Yes   
4   1665  Female  79.0             1              0          Yes   
5  56669    Male  81.0             0              0          Yes   
6  53882    Male  74.0             1              1          Yes   
7  10434  Female  69.0             0              0           No   
8  27419  Female  59.0             0              0          Yes   
9  60491  Female  78.0             0              0          Yes   

       work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
0        Private          Urban             228.69  36.6  formerly smoked   
1  Self-employed          Rural             202.21   NaN     never smoked   
2        Private          Rural             105.92  32.5     never smoked   
3        Private          Urban             171.23  34.4           smokes   
4  Self-employed          Rural             174.12  24.0     never smoked   
5        Private          Urban             186.21  29.0  formerly smoked   
6        Private          Rural              70.09  27.4     never smoked   
7        Private          Urban              94.39  22.8     never smoked   
8        Private          Rural              76.15   NaN          Unknown   
9        Private          Urban              58.57  24.2          Unknown   

   stroke  
0       1  
1       1  
2       1  
3       1  
4       1  
5       1  
6       1  
7       1  
8       1  
9       1

## Membagi Dataset menjadi Training Set dan Testing Set dengan Proporsi 70:30

In [37]:
X = dataset_split.iloc[:, :-1]
Y = dataset_split.iloc[:, -1]

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2)

In [38]:
print("Dimensi X_Train: ", X_train.shape)
print("Dimensi X_Test: ", X_test.shape)
print("Dimensi Y_Train: ", Y_train.shape)
print("Dimensi Y_Test: ", Y_test.shape)

Dimensi X_Train:  (4088, 11)
Dimensi X_Test:  (1022, 11)
Dimensi Y_Train:  (4088,)
Dimensi Y_Test:  (1022,)


# Data Cleaning

In [77]:
dataset_split = pd.read_csv("healthcare-dataset-stroke-data-rusak.csv")

In [78]:
dataset_split[22:33]

id  gender   age  hypertension  heart_disease ever_married  \
22  68794  Female  79.0             0              0          Yes   
23  64778    Male  82.0             0              1          Yes   
24   4219    Male  71.0             0              0          Yes   
25  70822    Male  80.0             0              0          Yes   
26  38047  Female  65.0             0              0          Yes   
27  61843    Male  58.0             0              0          Yes   
28  54827    Male  69.0             0              1          Yes   
29  69160    Male  59.0             0              0          Yes   
30  43717    Male  57.0             1              0          Yes   
31  33879    Male  42.0             0              0          Yes   
32  39373  Female  82.0             1              0          Yes   

        work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
22  Self-employed          Urban             228.70  26.6     never smoked   
23        Private          Rural             208.30  32.5          Unknown   
24        Private          Urban             102.87  27.2  formerly smoked   
25  Self-employed          Rural             104.12  23.5     never smoked   
26        Private          Rural             100.98  28.2  formerly smoked   
27        Private          Rural             189.84   NaN          Unknown   
28  Self-employed          Urban             195.23  28.3           smokes   
29        Private          Rural             211.78   NaN  formerly smoked   
30        Private          Urban             212.08  44.2           smokes   
31        Private          Rural              83.41  25.4          Unknown   
32  Self-employed          Urban             196.92  22.2     never smoked   

    stroke  
22       1  
23       1  
24       1  
25       1  
26       1  
27       1  
28       1  
29       1  
30       1  
31       1  
32       1

### Menangani Nilai NULL
#### Kita mengecek informasi ada atau tidaknya data null didalam value attribut dengan menggunakan dataset_cleaning.isna().sum(), dan terlihat pada tampilan dibawah bahwa atribut BMI memiliki 201 dataset yang bernilai null

In [189]:
dataset_split.isna().sum()

id                     0
gender                 0
age                    0
hypertension           0
heart_disease          0
ever_married           0
work_type              0
Residence_type         0
avg_glucose_level      0
bmi                  199
smoking_status         0
stroke                 0
dtype: int64

#### Mengubah Nilai NULL dengan Strategy Mean

In [86]:
from sklearn.impute import SimpleImputer

In [87]:
imputer = SimpleImputer(strategy="mean")  

In [88]:
dataset_split["bmi"] = imputer.fit_transform(dataset_split[["bmi"]])

In [89]:
dataset_split.isna().sum() ## Nilai Null pada Atribut BMI sudah diatasi

id                   0
gender               0
age                  0
hypertension         0
heart_disease        0
ever_married         0
work_type            0
Residence_type       0
avg_glucose_level    0
bmi                  0
smoking_status       0
stroke               0
dtype: int64

### Menangani Nilai Duplicate

In [97]:
dataset_split = pd.read_csv("healthcare-dataset-stroke-data-rusak.csv")

#### Mengecek Nilai Duplicate

In [98]:
dataset_split[dataset_split.duplicated()]

id gender   age  hypertension  heart_disease ever_married work_type  \
1  9046   Male  67.0             0              1          Yes   Private   
2  9046   Male  67.0             0              1          Yes   Private   
3  9046   Male  67.0             0              1          Yes   Private   
4  9046   Male  67.0             0              1          Yes   Private   
5  9046   Male  67.0             0              1          Yes   Private   
6  9046   Male  67.0             0              1          Yes   Private   
7  9046   Male  67.0             0              1          Yes   Private   
8  9046   Male  67.0             0              1          Yes   Private   
9  9046   Male  67.0             0              1          Yes   Private   

  Residence_type  avg_glucose_level   bmi   smoking_status  stroke  
1          Urban             228.69  36.6  formerly smoked       1  
2          Urban             228.69  36.6  formerly smoked       1  
3          Urban             228.69  36.6  formerly smoked       1  
4          Urban             228.69  36.6  formerly smoked       1  
5          Urban             228.69  36.6  formerly smoked       1  
6          Urban             228.69  36.6  formerly smoked       1  
7          Urban             228.69  36.6  formerly smoked       1  
8          Urban             228.69  36.6  formerly smoked       1  
9          Urban             228.69  36.6  formerly smoked       1

#### Mengecek Jumlah Nilai Duplicate

In [99]:
dataset_split.duplicated().sum()

9

#### Menangani Nilai Duplicate dengan drop_duplicates(inplace=True)

In [100]:
dataset_split.drop_duplicates(inplace=True)

#### Nilai Duplicate Berhasil ditangani

In [101]:
dataset_split[dataset_cleaning.duplicated()]

Empty DataFrame
Columns: [id, gender, age, hypertension, heart_disease, ever_married, work_type, Residence_type, avg_glucose_level, bmi, smoking_status, stroke]
Index: []

In [95]:
dataset_split.duplicated().sum()

0

## Data Transforming (Normalization)
### Melakukan Nomarlisasi setelah Cleaning Data

In [103]:
from sklearn.preprocessing import MinMaxScaler

In [104]:
min_max_scaler = MinMaxScaler()

In [110]:
x_scaled = min_max_scaler.fit_transform(
    dataset_split[
            ['avg_glucose_level', 
             'hypertension',
             'age']]
)

In [111]:
data_normalization = pd.DataFrame(x_scaled)
data_normalization.head()

0    1         2
0  0.801265  0.0  0.816895
1  0.116841  1.0  0.987793
2  0.301634  0.0  0.743652
3  0.228003  0.0  0.658203
4  0.760410  0.0  0.951172

## Data Transforming (Standarization)
### Melakukan Standarisasi setelah Cleaning Data

In [113]:
np.std(dataset_split)

id                   21159.420666
age                     22.597912
hypertension             0.296279
heart_disease            0.225454
avg_glucose_level       45.256067
bmi                      7.857202
stroke                   0.211745
dtype: float64

In [117]:
from sklearn.preprocessing import StandardScaler

In [118]:
standar_scaler = StandardScaler()

In [120]:
x_scaled2 = standar_scaler.fit_transform(
    dataset_split[
            ['avg_glucose_level', 
             'hypertension',
             'age']]
)

In [121]:
data_standarization = pd.DataFrame(x_scaled2)
data_standarization.head()

0         1         2
0  2.708550 -0.328191  1.054107
1 -0.567475  3.047011  1.673634
2  0.317047 -0.328191  0.788596
3 -0.035392 -0.328191  0.478833
4  2.512996 -0.328191  1.540878

In [122]:
np.std(x_scaled2)

1.0

In [123]:
np.mean(x_scaled2)

-1.2072215336706924e-17

## Mengubah Tipe Data Atribut Angka(Int) ke Object

In [159]:
### Tipe Data Seblum dirubah

In [167]:
tipe_angka = pd.read_csv("healthcare-dataset-stroke-data.csv")

In [180]:
tipe_angka["stroke"].dtypes

dtype('int64')

In [ ]:
### Tipe Data Sesudah Dirubah

In [181]:
tipe_objek = tipe_angka[["stroke"]].astype("object")
tipe_objek["stroke"].dtypes

dtype('O')

## One Hot Encoding pada Dataset

In [183]:
from sklearn.preprocessing import OneHotEncoder

In [184]:
dataset_oneHot = pd.read_csv("healthcare-dataset-stroke-data.csv")

In [185]:
one_hot_encoding = OneHotEncoder(sparse=False)

In [186]:
statusSmoking_encoded = one_hot_encoding.fit_transform(dataset_oneHot[['smoking_status']])

In [187]:
statusSmoking_encoded2 = pd.DataFrame(statusSmoking_encoded)
dataset_oneHot = dataset_oneHot.join(statusSmoking_encoded2 )

In [188]:
dataset_oneHot.head(10)

id  gender   age  hypertension  heart_disease ever_married  \
0   9046    Male  67.0             0              1          Yes   
1  51676  Female  61.0             0              0          Yes   
2  31112    Male  80.0             0              1          Yes   
3  60182  Female  49.0             0              0          Yes   
4   1665  Female  79.0             1              0          Yes   
5  56669    Male  81.0             0              0          Yes   
6  53882    Male  74.0             1              1          Yes   
7  10434  Female  69.0             0              0           No   
8  27419  Female  59.0             0              0          Yes   
9  60491  Female  78.0             0              0          Yes   

       work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
0        Private          Urban             228.69  36.6  formerly smoked   
1  Self-employed          Rural             202.21   NaN     never smoked   
2        Private          Rural             105.92  32.5     never smoked   
3        Private          Urban             171.23  34.4           smokes   
4  Self-employed          Rural             174.12  24.0     never smoked   
5        Private          Urban             186.21  29.0  formerly smoked   
6        Private          Rural              70.09  27.4     never smoked   
7        Private          Urban              94.39  22.8     never smoked   
8        Private          Rural              76.15   NaN          Unknown   
9        Private          Urban              58.57  24.2          Unknown   

   stroke    0    1    2    3  
0       1  0.0  1.0  0.0  0.0  
1       1  0.0  0.0  1.0  0.0  
2       1  0.0  0.0  1.0  0.0  
3       1  0.0  0.0  0.0  1.0  
4       1  0.0  0.0  1.0  0.0  
5       1  0.0  1.0  0.0  0.0  
6       1  0.0  0.0  1.0  0.0  
7       1  0.0  0.0  1.0  0.0  
8       1  1.0  0.0  0.0  0.0  
9       1  1.0  0.0  0.0  0.0